### PR TITLE
feat: document wrapped React components

### PR DESCRIPTION
**Pull request by mattrothenberg**
### Link to relevant issue

(Ignore the huge diff – this PR will be rebased once #300 is merged)

This is a companion PR to https://github.com/microsoft/vscode-webview-ui-toolkit/pull/300. It adds a nested `react-storybook` directory to the root of this repository, inside of which lives a standalone [Storybook](https://storybook.js.org/)  instance that documents all of the wrapped React components.

### Description of changes

It would be fantastic to leverage Storybook's [composition feature](https://storybook.js.org/docs/react/sharing/storybook-composition) to document both the web components and wrapped React components. (cc @hawkticehurst per our conversation on Slack)

To that end, I've added a standalone Storybook to this repository that you can interact with by `cd`ing to the directory and running:

```bash
npm install
npm run storybook
```

I was able to document _most_ of the components without trouble, though I still ran into some show-stopping issues. Unclear whether it's user error (read: me), or an outstanding issue with the FAST wrapper. Here's a rough list:

## VSCodeButton
- [ ]  `appearance` prop doesn't seem to be working
- [x] `onClick` events seem to be propagating even when the button is `disabled`
- [ ] `autofocus` prop doesn't seem to be working

## VSCodeCheckbox
- [x] `onChange` event doesn't seem to be firing (**fixed** pending a fast wrapper update OR us manually registering these events on the wrapped component)

## VSCodeDropdown
- [ ] `position` prop doesn't seem to be working
- [x] `onChange` event doesn't seem to be firing (**fixed** pending a fast wrapper update OR us manually registering these events on the wrapped component)

## VSCodePanels
- [ ] `activeid` prop isn't camel-cased (and should be IMO)
- [x] `onChange` event doesn't seem to be firing (**fixed** pending a fast wrapper update OR us manually registering these events on the wrapped component)

## VSCodeTextField
- [x] `onChange` event doesn't seem to be firing (**fixed** pending a fast wrapper update OR us manually registering these events on the wrapped component)

## VSCodeRadio
- [x] `onChange` event doesn't seem to be firing (**fixed** pending a fast wrapper update OR us manually registering these events on the wrapped component)

## VSCodeRadioGroup
- [x] `onChange` event doesn't seem to be firing (**fixed** pending a fast wrapper update OR us manually registering these events on the wrapped component)

### CODE REVIEW

**Comment by hawkticehurst** [comment]
@mattrothenberg This looks great thus far!! 

I installed the changes locally and am cross-referencing against your notes now. Off the bat, I discovered that all the onChange errors seem to be a simple case of misconfiguring Storybook.

When defining the `argTypes`, you've used the action keyword `'change'` in a few places, when instead it should be `'changed'`.

```ts
argTypes: {
  onChange: {action: 'changed'}, // keyword: "changed" not "change"
}
```

Also when later defining the `onChange` event in the template you have to pair it with the `action` function from the `@storybook/addon-actions` package.

```ts
import {action} from '@storybook/addon-actions'; // import action function

... other code ...

const Template = args => {
const {label, onChange, ...rest} = args;
  return (
    <>
      <VSCodeCheckbox onChange={action('checkbox-onchange')} {...rest}> // explicitly use action function in template
        {label}
      </VSCodeCheckbox>
    </>
  );
};
```

_(As a minor side note/question, is there a particular reason you're wrapping the Checkbox component in a React Fragment?)_

I'm still working through the other issues you listed, but in the meantime, this should hopefully resolve the vast majority of issues!

**Comment by mattrothenberg** [comment]
@hawkticehurst oh wow, great catch! are you able to push your changes to this branch by any chance?

No reason why I wrapped the Checkbox in a fragment :)

**Comment by mattrothenberg** [comment]
@hawkticehurst Adding the custom event to the React component is proving to be the solution here, thanks for the advice!

For clarification's sake, I don't _think_ this is strictly true:

> When defining the argTypes, you've used the action keyword 'change' in a few places, when instead it should be 'changed'.

I could def be wrong, but from my tinkering, the string value you pass here simply dictates what shows up in the "Actions" panel in Storybook. 

```tsx
onChange: {action: 'barbaz'} // this is valid
onChange: {action: 'foo'} // this is valid,
onChange: {action: 'change'} // this is valid
```

Screen Shot 2022-01-11 at 8 52 25 AM

Did you want to use `changed` for reasons of keeping our Storybook documentation consistent?

**Comment by hawkticehurst** [comment]
> @hawkticehurst Adding the custom event to the React component is proving to be the solution here, thanks for the advice!
> 
> For clarification's sake, I don't _think_ this is strictly true:
> 
> > When defining the argTypes, you've used the action keyword 'change' in a few places, when instead it should be 'changed'.
> 
> I could def be wrong, but from my tinkering, the string value you pass here simply dictates what shows up in the "Actions" panel in Storybook.
> 
> ```tsx
> onChange: {action: 'barbaz'} // this is valid
> onChange: {action: 'foo'} // this is valid,
> onChange: {action: 'change'} // this is valid
> ```
> 
> Screen Shot 2022-01-11 at 8 52 25 AM
> 
> Did you want to use `changed` for reasons of keeping our Storybook documentation consistent?

Ah! Thanks for the call out, looking back at the docs there does indeed seem to be a lot of flexibility in the naming here!

Also, given this flexibility, I think it makes sense to just use whatever the HTML change event is named––so I double checked and HTML actually does use `change` (not `changed`) 😅.

So I'm happy to leave everything as `change`.

File: react-storybook/stories/TextArea.stories.tsx [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+	const {label, ...rest} = args;
+	return (
+		<>
+			<VSCodeTextArea {...rest}>{label}</VSCodeTextArea>
+		</>
+	);
+};
+
+export const Default = Template.bind({});
+Default.args = {};
```
Comment (hawkticehurst) [inline react-storybook/stories/TextArea.stories.tsx:65]: Set default values and use spread operator elsewhere

File: react-storybook/stories/Tag.stories.tsx [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+	const {label, ...rest} = args;
+	return (
+		<>
+			<VSCodeTag {...rest}>{label}</VSCodeTag>
+		</>
+	);
+};
+
+export const Default = Template.bind({});
+Default.args = {};
```
Comment (hawkticehurst) [inline react-storybook/stories/Tag.stories.tsx:26]: Set default values based on web component story

File: react-storybook/stories/RadioGroup.stories.tsx [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+	},
+};
+
+const Template = args => {
+	const {label, numberOfChildren, ...rest} = args;
+
+	const options = Array.from({length: numberOfChildren}, (_, i) => i);
+
+	return (
+		<>
```
Comment (hawkticehurst) [inline react-storybook/stories/RadioGroup.stories.tsx:30]: Remove fragment

File: react-storybook/stories/RadioGroup.stories.tsx [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+						{`Option #${index + 1}`}
+					</VSCodeRadio>
+				))}
+			</VSCodeRadioGroup>
+		</>
+	);
+};
+
+export const Default = Template.bind({});
+Default.args = {};
```
Comment (hawkticehurst) [inline react-storybook/stories/RadioGroup.stories.tsx:44]: Set default values and use spread operator elsewhere

File: react-storybook/stories/Radio.stories.tsx [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+		value: {
+			control: 'text',
+		},
+	},
+};
+
+const Template = args => {
+	const {label, ...rest} = args;
+	return (
+		<>
```
Comment (hawkticehurst) [inline react-storybook/stories/Radio.stories.tsx:35]: Remove fragment

File: react-storybook/stories/ProgressRing.stories.tsx [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+import {VSCodeProgressRing} from '../../src/react';
+
+export default {
+	title: 'Library/Progress Ring',
+	component: VSCodeProgressRing,
+};
+
+const Template = () => {
+	return (
+		<>
```
Comment (hawkticehurst) [inline react-storybook/stories/ProgressRing.stories.tsx:12]: Remove fragment

File: react-storybook/stories/ProgressRing.stories.tsx [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+			<VSCodeProgressRing />
+		</>
+	);
+};
+
+export const Default = Template.bind({});
+Default.args = {};
```
Comment (hawkticehurst) [inline react-storybook/stories/ProgressRing.stories.tsx:19]: Set default values based on web component story

File: react-storybook/stories/Option.stories.tsx [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+			defaultValue: false,
+			control: 'boolean',
+		},
+	},
+};
+
+const Template = args => {
+	const {label, ...rest} = args;
+	return (
+		<>
```
Comment (hawkticehurst) [inline react-storybook/stories/Option.stories.tsx:27]: Remove fragment

File: react-storybook/stories/Option.stories.tsx [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+			<VSCodeOption {...rest}>{label}</VSCodeOption>
+		</>
+	);
+};
+
+export const Default = Template.bind({});
+Default.args = {};
```
Comment (hawkticehurst) [inline react-storybook/stories/Option.stories.tsx:34]: Set default values and use spread operator elsewhere

File: react-storybook/stories/Link.stories.tsx [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+			defaultValue:
+				'https://github.com/microsoft/vscode-webview-ui-toolkit',
+		},
+	},
+};
+
+const Template = args => {
+	const {label, ...rest} = args;
+	return (
+		<>
```
Comment (hawkticehurst) [inline react-storybook/stories/Link.stories.tsx:24]: Remove fragment

File: react-storybook/stories/Link.stories.tsx [hunk 11]
```diff
@@ -0,0 +1,31 @@
...
+			<VSCodeLink {...rest}>{label}</VSCodeLink>
+		</>
+	);
+};
+
+export const Default = Template.bind({});
+Default.args = {};
```
Comment (hawkticehurst) [inline react-storybook/stories/Link.stories.tsx:31]: Set default values and use spread operator elsewhere

File: react-storybook/stories/Dropdown.stories.tsx [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+						Option #{index + 1}
+					</VSCodeOption>
+				))}
+			</VSCodeDropdown>
+		</div>
+	);
+};
+
+export const Default = Template.bind({});
+Default.args = {};
```
Comment (hawkticehurst) [inline react-storybook/stories/Dropdown.stories.tsx:63]: Set default values and use spread operator elsewhere

File: react-storybook/stories/Divider.stories.tsx [hunk 13]
```diff
@@ -0,0 +1,27 @@
...
+		},
+	},
+};
+
+const Template = args => {
+	return <VSCodeDivider {...args} />;
+};
+
+export const Default = Template.bind({});
+Default.args = {};
```
Comment (hawkticehurst) [inline react-storybook/stories/Divider.stories.tsx:22]: Set default values and use spread operator elsewhere

File: react-storybook/stories/DataGrid.stories.tsx [hunk 14]
```diff
@@ -0,0 +1,86 @@
...
+						))}
+					</VSCodeDataGridRow>
+				);
+			})}
+		</VSCodeDataGrid>
+	);
+};
+
+export const Default = Template.bind({});
+Default.args = {};
```
Comment (hawkticehurst) [inline react-storybook/stories/DataGrid.stories.tsx:86]: Set default values based on web component story

**Review by hawkticehurst — COMMENTED** [review]
Gave an initial pass on this work and it's looking good! I left a handful of comments mainly related to matching the conventions that are established in the web component stories.

Let me know if you have any questions!

**Comment by hawkticehurst** [comment]
I also noticed that Storybook doesn't seem to be correctly auto-generating code snippets. I'm not sure if this is a result of these really being web components under the hood, but if you could explore how to resolve this issue that would be appreciated! 🙂

P.S. If you can't figure out what the problem is (or if it's unresolvable) you can default to manually defining code snippets like we do in the web components ([example](https://github.com/microsoft/vscode-webview-ui-toolkit/blob/main/src/button/button.stories.ts#L45-L51)).

Screen Shot 2022-01-12 at 10 25 16 AM
Screen Shot 2022-01-12 at 10 25 28 AM

File: react-storybook/stories/Button.stories.tsx [hunk 15]
```diff
@@ -0,0 +1,99 @@
...
+Default.args = {
+	label: 'Button Text',
+	appearance: 'primary',
+};
+
+export const Secondary = Template.bind({});
+Secondary.args = {
+	label: 'Button Text',
+	appearance: 'secondary',
+};
```
[the remark this answers is not in the document]
Comment (mattrothenberg) [reply react-storybook/stories/Button.stories.tsx:72]: @hawkticehurst I definitely can do, though I'm curious why you prefer this over setting default values in the `argTypes` object up top.

**Comment by mattrothenberg** [comment]
@hawkticehurst re: the `[object Object]` display issue, I think that's a bug in the React wrapper.

I noticed that this resolves the issue, though.

```tsx
import {VSCodeBadge} from '../../src/react';
VSCodeBadge.displayName = 'VSCodeBadge';

export default {
	title: 'Library/Badge',
	component: VSCodeBadge,
	argTypes: {
		label: {
			control: 'text',
		},
	},
};
// etc.
```

File: react-storybook/stories/Button.stories.tsx (hunk 15 above)
[the remark this answers is not in the document]
Comment (mattrothenberg) [reply react-storybook/stories/Button.stories.tsx]: Working through making these changes and it feels a little painful – setting `defaultValue` in the arg types definition effectively (and much more simply) provides this functionality, though I'd concede that it's definitely more implicit than manually defining the arg types on the default story. WDYT?

[the remark this answers is not in the document]
Comment (hawkticehurst) [reply react-storybook/stories/Button.stories.tsx]: Ahh I see, interesting! It is indeed so implicit I missed on the first pass.

Regardless, I agree that we can roll with the easier version and you can revert the initial work you've done on converting to be more explicit (sorry about that 😅).

[the remark this answers is not in the document]
Comment (hawkticehurst) [reply react-storybook/stories/Button.stories.tsx]: I'll also resolve all the other conversations related to this topic.

**Comment by mattrothenberg** [comment]
Running into spooky problems with this branch. Closing and will re-open with another PR. I'll copy over all relevant context.

**Pull request closed without merging**